### ISSUE TITLE
学习 DeepSeek (Part 13.2: nn.Embedding)

### ISSUE THREAD

**Issue by nonocast** [nonocast/me#353]
从NLP的发展来看, 文本embedding经历过程如下:
1. 词袋法 (1990年代)
2. TF-IDF (2000年代)
3. Word2vec (2013)
4. BGE (2023)

根据输入的颗粒度分为: 词向量 (Word Embedding), 句向量 (Sentence Embedding) 和 文档向量 (Document Embedding)，也就是说Word2vec和BGE是具有语义的，但Word2vec是静态语义，而基于transformer/BERT的BGE则是动态语义。

我们知道embedding是通过一个vector来表达一个text, 这个vector你可以理解为是特征，即维度，记为d，Word2vec因为是Word Embedding，所以如果输入3xd, 但是如果采用BGE，因为是Sentence Embedding，则输出的是1xd，一定先把输入输出搞明白。

# nn.Embedding
我们从最最简单的nn.Embedding开始，

main.py
```py
import torch
import torch.nn as nn

word_to_ix = {"I": 0, "love": 1, "NLP": 2}
ix_to_word = {v: k for k, v in word_to_ix.items()}

# 返回一个3x5的随机初始化词向量矩阵
embedding = nn.Embedding(num_embeddings=len(word_to_ix), embedding_dim=5)
print(embedding.weight.detach())

input_ids = torch.tensor([0, 1, 2], dtype=torch.long)
embedded = embedding(input_ids)

print("\n词向量表示（逐词）：")
for i, vec in enumerate(embedded):
    values = [f"{x.item():.4f}" for x in vec]
    print(f"{ix_to_word[input_ids[i].item()]} → {values}")
```

代码说明:
1. nn.Embedding构建一个3x5的随机表格
2. embedding就是查表，根据id去表格提取向量

```sh
❯ uv init
❯ uv pip install torch numpy
❯ uv run main.py
tensor([[-0.1877,  1.7682, -0.8652,  1.4148, -0.3481],
        [-1.9682,  2.6889, -1.4339,  0.3031, -0.8815],
        [-0.2949,  1.9599, -1.2254, -0.6882,  0.3846]])

词向量表示（逐词）：
I → ['-0.1877', '1.7682', '-0.8652', '1.4148', '-0.3481']
love → ['-1.9682', '2.6889', '-1.4339', '0.3031', '-0.8815']
NLP → ['-0.2949', '1.9599', '-1.2254', '-0.6882', '0.3846']
```

到目前为止，我们只是将word嵌入到随机表格中，但是没有任何意义。

main.py
```py
import torch
import torch.nn as nn
import torch.nn.functional as F

# 词表映射
word_to_ix = {
    "I": 0, "love": 1, "NLP": 2,
    "hate": 3, "bugs": 4,
    "enjoy": 5, "ML": 6,
    "dislike": 7, "errors": 8,
}
ix_to_word = {v: k for k, v in word_to_ix.items()}

# 训练数据
training_data = [
    ([0, 1, 2], 1),   # I love NLP → 正面
    ([0, 3, 4], 0),   # I hate bugs → 负面
    ([0, 5, 6], 1),   # I enjoy ML → 正面
    ([0, 7, 8], 0),   # I dislike errors → 负面
]

# 模型定义
class SentimentClassifier(nn.Module):
    def __init__(self, vocab_size, embedding_dim, hidden_dim):
        super().__init__()
        self.embedding = nn.Embedding(vocab_size, embedding_dim)
        self.linear = nn.Linear(embedding_dim, hidden_dim)
        self.output = nn.Linear(hidden_dim, 2)

    def forward(self, input_ids):
        embeds = self.embedding(input_ids)
        sentence_vec = embeds.mean(dim=0)
        hidden = F.relu(self.linear(sentence_vec))
        logits = self.output(hidden)
        return logits

# 打印 embedding 的格式化函数
def print_embedding_matrix(embedding, word_to_ix, decimals=4):
    print("\n词向量矩阵：")
    weight = embedding.weight.detach()
    ix_to_word = {v: k for k, v in word_to_ix.items()}
    for idx, vec in enumerate(weight):
        word = ix_to_word.get(idx, f"id={idx}")
        formatted = "  ".join(f"{x.item(): .{decimals}f}" for x in vec)
        print(f"{word:<10} →  {formatted}")

# 超参数
EMBED_DIM = 10
HIDDEN_DIM = 8
EPOCHS = 30

# 模型初始化
model = SentimentClassifier(vocab_size=len(word_to_ix), embedding_dim=EMBED_DIM, hidden_dim=HIDDEN_DIM)

# 训练前打印 embedding
print("Before training:")
print_embedding_matrix(model.embedding, word_to_ix)

# 优化器和损失函数
optimizer = torch.optim.Adam(model.parameters(), lr=0.01)
loss_fn = nn.CrossEntropyLoss()

# 训练
for epoch in range(EPOCHS):
    total_loss = 0.0
    for sentence, label in training_data:
        input_tensor = torch.tensor(sentence, dtype=torch.long)
        target = torch.tensor([label], dtype=torch.long)

        optimizer.zero_grad()
        logits = model(input_tensor)
        loss = loss_fn(logits.unsqueeze(0), target)
        loss.backward()
        optimizer.step()

        total_loss += loss.item()

    if (epoch + 1) % 5 == 0 or epoch == 0:
        print(f"Epoch {epoch+1}: Loss = {total_loss:.4f}")

# 训练后打印 embedding
print("\nAfter training:")
print_embedding_matrix(model.embedding, word_to_ix)

# 预测函数
def predict(sentence):
    ids = torch.tensor([word_to_ix[w] for w in sentence], dtype=torch.long)
    logits = model(ids)
    probs = F.softmax(logits, dim=-1)
    label = torch.argmax(probs).item()
    return "正面" if label == 1 else "负面", probs.detach().numpy()

# 测试预测
print("\n测试结果：")
for test in [["I", "love", "NLP"], ["I", "hate", "bugs"], ["I", "enjoy", "ML"], ["I", "dislike", "errors"]]:
    result, prob = predict(test)
    print(f"{' '.join(test):<25} → {result}  (概率 = {prob.round(3)})")
```

运行后输出如下:
```sh
❯ make
Before training:

词向量矩阵：
I          →   0.4626  -0.8508   0.1222  -0.3242   1.3655   1.1354  -0.0558  -1.3324   1.0001  -0.8592
love       →   1.2108   1.4096   0.1097   0.2584  -0.0804  -0.1645   2.5466  -0.8665   0.3521  -0.2048
NLP        →  -1.0624   0.0101   1.9710   1.9706   0.5750  -0.5945  -1.3419  -0.4447   0.3548  -2.0006
hate       →  -0.1867   0.9188  -0.6163  -0.7524   0.7755  -0.8337  -0.0852  -0.0098   1.0922   0.4019
bugs       →   1.2237  -1.4327   1.0232   1.1406  -0.9049  -0.9573  -0.0239  -0.6844   0.8589  -2.4124
enjoy      →   0.1292   0.1043   0.3356   1.0830  -1.5035  -0.9080   1.5711  -0.0567  -0.1690  -1.7818
ML         →   0.1460   0.2101  -0.1325  -0.9411   1.0088  -0.3902  -0.3596  -2.7933   0.9246  -0.1109
dislike    →  -1.6319   0.3736  -0.6279  -0.1869   1.3065  -0.3849  -0.8728   0.6059  -0.7109  -0.5723
errors     →   0.3426   1.2024  -0.4775   0.2119   1.6263  -1.8830  -1.0280  -2.2962  -0.8988   1.0015
Epoch 1: Loss = 2.7494
Epoch 5: Loss = 2.4328
Epoch 10: Loss = 1.7091
Epoch 15: Loss = 0.6531
Epoch 20: Loss = 0.1110
Epoch 25: Loss = 0.0359
Epoch 30: Loss = 0.0200

After training:

词向量矩阵：
I          →   0.2101  -0.5891  -0.0992  -0.1884   1.6533   0.8942  -0.2522  -1.7147   1.0068  -0.7667
love       →   1.6669   1.0287   0.2926   0.7828  -0.4357   0.3338   3.0680  -1.3018   0.5530  -0.4401
NLP        →  -0.6063  -0.3708   2.1538   2.4950   0.2197  -0.0962  -0.8206  -0.8800   0.5557  -2.2360
hate       →  -0.7418   1.4523  -0.5940  -1.2921   1.3337  -1.3960  -0.6434   0.3592   1.4603   0.7640
bugs       →   0.6686  -0.8993   1.0454   0.6009  -0.3467  -1.5197  -0.5822  -0.3154   1.2269  -2.0504
enjoy      →   0.5695  -0.2686   0.5337   1.6312  -1.8391  -0.4294   2.0228  -0.5050  -0.1701  -2.1191
ML         →   0.5863  -0.1629   0.0656  -0.3928   0.6732   0.0884   0.0921  -3.2416   0.9235  -0.4482
dislike    →  -2.0240   0.7341  -0.9595  -0.2678   1.7101  -0.7578  -1.2656   0.4899  -0.9881  -0.4174
errors     →  -0.0496   1.5629  -0.8091   0.1311   2.0298  -2.2560  -1.4208  -2.4122  -1.1759   1.1564

测试结果：
I love NLP                → 正面  (概率 = [0.002 0.998])
I hate bugs               → 负面  (概率 = [0.989 0.011])
I enjoy ML                → 正面  (概率 = [0.004 0.996])
I dislike errors          → 负面  (概率 = [0.999 0.001])
```

代码说明:
1. SentimentClassifier 是一个基于神经网络的文本情感分类模型。
2. 模型结构采用 FNN（Feedforward Neural Network，前馈神经网络），整体由三层组成：
	1. 输入层：将词 ID 映射为词向量（Embedding），通过 Mean Pooling 得到一个固定长度的句向量（10维，EMBED_DIM=10）。
	2. 隐藏层：使用全连接层将句向量变换为 8 维特征表示，并应用 ReLU 激活函数：Linear(10 → 8) + ReLU。
	3. 输出层：通过全连接层输出两个维度的 logits，表示句子属于“正面”或“负面”的预测分数：Linear(8 → 2)。
3. 通过训练过程（反向传播 + 梯度下降），模型学习各层的参数（包括词向量和全连接层权重），以最小化预测结果与真实标签之间的交叉熵损失，从而提升情感分类准确率。

这样就形成了一个特定网络模型用来识别情感。这里只是用4句作为样本，实际使用需要至少几百条甚至上万条语句。

参考阅读:
- [大模型靠啥理解文字？通俗解释：词嵌入embedding_哔哩哔哩_bilibili](https://www.bilibili.com/video/BV1bfoQYCEHC?spm_id_from=333.788.recommend_more_video.-1&vd_source=01aee54449fa414da59ad179ade9d884)
- [从词袋法到词嵌入，词向量是如何演进的？ BGE | TF-IDF | 词袋法 | WordEmbedding_哔哩哔哩_bilibili](https://www.bilibili.com/video/BV1CWV6zKEBn/?spm_id_from=333.337.search-card.all.click&vd_source=01aee54449fa414da59ad179ade9d884)
